src/parakeet/Driver.cpp

```diff
@@ -0,0 +1,124 @@
+/*
+	Copyright 2021 OpenJAUS, LLC (dba MechaSpin). Subject to the MIT license.
+*/
+
+#include <parakeet/Driver.h>
+#include <parakeet/exceptions/NotConnectedToSensorException.h>
+
+#include <iostream>
+
+namespace mechaspin
+{
+namespace parakeet
+{
+	void Driver::stop()
+	{
+        runUpdateThread = false;
+
+        if(updateThread.joinable())
+        {
+            updateThread.join();
+        }
+	}
+
+	void Driver::close()
+	{
+        bool isc = isConnected();
+        bool runUpdate = runUpdateThread;
+        bool isJoin = updateThread.joinable();
+        if (isConnected() || runUpdateThread || updateThread.joinable())
+        {
+           stop();
+        }
+	}
+
+    void Driver::start()
+    {
+        throwExceptionIfNotConnected();
+
+        runUpdateThread = true;
+        updateThreadStartTime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch());
+        updateThreadFrameCount = 0;
+        updateThread = std::thread([&] { this->updateThreadMainLoop(); });
+    }
+
+	void Driver::updateThreadMainLoop()
+	{
+        while (runUpdateThread)
+        {
+			if(updateThreadCallbackFunction)
+			{
+				updateThreadCallbackFunction();
+			}
+		}
+	}
+
+	void Driver::throwExceptionIfNotConnected()
+	{
+        if(!isConnected())
+        {
+            throw exceptions::NotConnectedToSensorException();
+        }
+	}
+
+    bool Driver::isUpdateThreadRunning()
+	{
+		return runUpdateThread;
+	}
+	
+
+    double Driver::getScanRate_Hz()
+    {
+        std::chrono::milliseconds currentTime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch());
+
+        auto ms = (currentTime - updateThreadStartTime).count();
+
+        if(ms == 0)
+        {
+            return 0;
+        }
+
+        return updateThreadFrameCount / (ms / 1000.0);
+    }
+
+    void Driver::registerScanCallback(std::function<void(const ScanDataPolar&)> callback)
+    {
+        scanCallbackFunction = callback;
+    }
+
+    void Driver::registerUpdateThreadCallback(std::function<void()> callback)
+    {
+        updateThreadCallbackFunction = callback;
+    }
+    
+    void Driver::onScanDataReceived(ScanData* scanData)
```

Comment:
It would be better to use a const ScanData& for the parameter. This prevents this method from accidentally modifying the scan data , which it shouldn't be doing, and maintains the responsibility lifetime management to the caller.

You transfer ownership here from the caller to this method which in unnecessary here since this method only really cares to use the ScanData, not what happens with it after it's done using it.

If you really need to transfer ownership I would suggest the use of the unique_ptr to help indicate that ownership has been transferred. As it is, it's ambiguous as to who owns the pointer after onScanDataReceived is called which could easily lead to memory leaks. 

See sections F and R here: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-conventional